File: loopx/control_plane/todos/completion_state.py
"""Python compatibility facade for TypeScript-owned Todo completion state."""

from __future__ import annotations

from enum import Enum
from functools import lru_cache
from typing import Any, Mapping

from ..effect_runtime import EffectRuntimeRejected, effect_runtime_result


TODO_COMPLETION_STATE_REQUEST_SCHEMA = "loopx_todo_completion_state_request_v0"
TODO_COMPLETION_STATE_RESULT_SCHEMA = "loopx_todo_completion_state_result_v0"


class TodoCompletionContinuation(str, Enum):
    """Stable Python import compatibility for the persisted wire values."""

    ACTIVE_GOAL = "active_goal"
    SUCCESSOR = "successor"
    NO_FOLLOWUP = "no_followup"


class TodoCompletionRecovery(str, Enum):
    """Stable Python import compatibility for the persisted wire value."""

    SAME_TURN_TERMINAL_CLOSEOUT = "same_turn_terminal_closeout"
    LIFECYCLE_REENTRY_TERMINAL_CLOSEOUT = "lifecycle_reentry_terminal_closeout"


def _string_value(value: Any) -> str:
    return str(value or "")


def _no_followup_value(value: Any) -> bool | str:
    return value if isinstance(value, bool) else _string_value(value)


def _result(method: str, params: dict[str, Any]) -> Mapping[str, Any]:
    try:
        result = effect_runtime_result(method, params)
    except EffectRuntimeRejected as exc:
        raise ValueError(str(exc)) from None
    if not isinstance(result, Mapping):
        raise RuntimeError("TypeScript Todo completion state result must be an object")
    if result.get("schema_version") != TODO_COMPLETION_STATE_RESULT_SCHEMA:
        raise RuntimeError("TypeScript Todo completion state result shape mismatch")
    return result


@lru_cache(maxsize=64)
def _normalize_cached(kind: str, value: bool | str) -> bool | str | None:
    result = _result(
        "todo.completion_state.normalize",
        {
            "schema_version": TODO_COMPLETION_STATE_REQUEST_SCHEMA,
            "kind": kind,
            "value": value,
        },
    )
    normalized = result.get("value")
    if kind == "no_followup":
        if normalized is not None and not isinstance(normalized, bool):
            raise RuntimeError("TypeScript no_followup normalization shape mismatch")
    elif normalized is not None and not isinstance(normalized, str):
        raise RuntimeError(
            "TypeScript completion metadata normalization shape mismatch"
        )
    return normalized


def normalize_todo_no_followup(value: Any) -> bool | None:
    normalized = _normalize_cached("no_followup", _no_followup_value(value))
    return normalized if isinstance(normalized, bool) else None


def normalize_todo_completion_continuation(value: Any) -> str | None:
    normalized = _normalize_cached("continuation", _string_value(value))
    return normalized if isinstance(normalized, str) else None


def require_todo_completion_continuation(value: Any) -> str:
    result = _result(
        "todo.completion_state.require_metadata",
        {
            "schema_version": TODO_COMPLETION_STATE_REQUEST_SCHEMA,
            "key": "completion_continuation",
            "value": _string_value(value),
        },
    )
    normalized = result.get("value")
    if not isinstance(normalized, str):
        raise RuntimeError("TypeScript completion continuation result shape mismatch")
    return normalized


def normalize_todo_completion_recovery(value: Any) -> str | None:
    normalized = _normalize_cached("recovery", _string_value(value))
    return normalized if isinstance(normalized, str) else None


def require_todo_completion_metadata(key: str, value: Any) -> str | None:
    result = _result(
        "todo.completion_state.require_metadata",
        {
            "schema_version": TODO_COMPLETION_STATE_REQUEST_SCHEMA,
            "key": key,
            "value": _string_value(value),
        },
    )
    normalized = result.get("value")
    if normalized is not None and not isinstance(normalized, str):
        raise RuntimeError("TypeScript completion metadata result shape mismatch")
    return normalized


def completion_continuation_for_write(*, no_followup: bool, has_successor: bool) -> str:
    result = _result(
        "todo.completion_state.continuation_for_write",
        {
            "schema_version": TODO_COMPLETION_STATE_REQUEST_SCHEMA,
            "no_followup": no_followup,
            "has_successor": has_successor,
        },
    )
    continuation = result.get("continuation")
    if continuation not in {item.value for item in TodoCompletionContinuation}:
        raise RuntimeError("TypeScript completion continuation result shape mismatch")
    return str(continuation)

File: loopx/control_plane/todos/completion_state.ts
import type { JsonObject } from "../effect_program.ts";
import {
  requireBoolean as requiredBoolean,
  requireJsonObject as requiredObject,
  requireStringArray as stringArray,
} from "../runtime_decode.ts";

export const TODO_COMPLETION_STATE_REQUEST_SCHEMA =
  "loopx_todo_completion_state_request_v0";
export const TODO_COMPLETION_STATE_RESULT_SCHEMA =
  "loopx_todo_completion_state_result_v0";

export const TODO_COMPLETION_CONTINUATIONS = [
  "active_goal",
  "successor",
  "no_followup",
] as const;
export const TODO_COMPLETION_RECOVERIES = [
  "same_turn_terminal_closeout",
  "lifecycle_reentry_terminal_closeout",
] as const;

export type TodoCompletionContinuation =
  (typeof TODO_COMPLETION_CONTINUATIONS)[number];
export type TodoCompletionRecovery =
  (typeof TODO_COMPLETION_RECOVERIES)[number];
export type TodoCompletionNormalizationKind =
  | "no_followup"
  | "continuation"
  | "recovery";

export interface TodoCompletionStateResult extends JsonObject {
  schema_version: typeof TODO_COMPLETION_STATE_RESULT_SCHEMA;
  continuation: TodoCompletionContinuation;
  recovery: TodoCompletionRecovery | null;
}

function optionalBoolean(value: unknown, label: string): boolean | null {
  if (value === null || value === undefined) return null;
  return requiredBoolean(value, label);
}

function optionalString(value: unknown, label: string): string | null {
  if (value === null || value === undefined) return null;
  if (typeof value !== "string") {
    throw new Error(`${label} must be a string or null`);
  }
  return value;
}

function requestObject(value: unknown, operation: string): JsonObject {
  const request = requiredObject(value, `todo.completion_state.${operation} params`);
  if (request.schema_version !== TODO_COMPLETION_STATE_REQUEST_SCHEMA) {
    throw new Error("Todo completion state request schema mismatch");
  }
  return request;
}

export function normalizeTodoNoFollowup(value: unknown): boolean | null {
  if (typeof value === "boolean") return value;
  if (value === null || value === undefined) return null;
  if (typeof value !== "string") {
    throw new Error("no_followup must be a boolean, string, or null");
  }
  const candidate = value.trim().toLowerCase();
  if (["1", "true", "yes", "y", "no_followup", "no-followup"].includes(candidate)) {
    return true;
  }
  if (["0", "false", "no", "n"].includes(candidate)) return false;
  return null;
}

export function normalizeTodoCompletionContinuation(
  value: unknown,
): TodoCompletionContinuation | null {
  if (value === null || value === undefined) return null;
  if (typeof value !== "string") {
    throw new Error("completion_continuation must be a string or null");
  }
  const candidate = value.trim().toLowerCase();
  return TODO_COMPLETION_CONTINUATIONS.some((item) => item === candidate)
    ? candidate as TodoCompletionContinuation
    : null;
}

export function requireTodoCompletionContinuation(
  value: unknown,
): TodoCompletionContinuation {
  const normalized = normalizeTodoCompletionContinuation(value);
  if (normalized !== null) return normalized;
  throw new Error(
    "completion_continuation must be one of: active_goal, successor, no_followup",
  );
}

export function normalizeTodoCompletionRecovery(
  value: unknown,
): TodoCompletionRecovery | null {
  if (value === null || value === undefined) return null;
  if (typeof value !== "string") {
    throw new Error("completion_recovery must be a string or null");
  }
  const candidate = value.trim().toLowerCase();
  return TODO_COMPLETION_RECOVERIES.some((item) => item === candidate)
    ? candidate as TodoCompletionRecovery
    : null;
}

export function requireTodoCompletionMetadata(
  key: string,
  value: unknown,
): TodoCompletionContinuation | TodoCompletionRecovery | null {
  if (key === "completion_continuation") {
    return requireTodoCompletionContinuation(value);
  }
  const normalized = normalizeTodoCompletionRecovery(value);
  if (normalized !== null || value === null || value === undefined || value === "") {
    return normalized;
  }
  throw new Error(
    "completion_recovery must be same_turn_terminal_closeout or " +
      "lifecycle_reentry_terminal_closeout",
  );
}

export function completionContinuationForWrite(
  noFollowup: boolean,
  hasSuccessor: boolean,
): TodoCompletionContinuation {
  if (noFollowup && hasSuccessor) {
    throw new Error("todo completion cannot record both no_followup and a successor");
  }
  if (noFollowup) return "no_followup";
  if (hasSuccessor) return "successor";
  return "active_goal";
}

export function normalizeTodoCompletionValue(value: unknown): JsonObject {
  const request = requestObject(value, "normalize");
  const kind = request.kind;
  if (kind !== "no_followup" && kind !== "continuation" && kind !== "recovery") {
    throw new Error("completion normalization kind is unsupported");
  }
  const raw = request.value;
  const normalized = kind === "no_followup"
    ? normalizeTodoNoFollowup(raw)
    : kind === "continuation"
    ? normalizeTodoCompletionContinuation(raw)
    : normalizeTodoCompletionRecovery(raw);
  return {
    schema_version: TODO_COMPLETION_STATE_RESULT_SCHEMA,
    kind,
    value: normalized,
  };
}

export function requireTodoCompletionMetadataValue(value: unknown): JsonObject {
  const request = requestObject(value, "require_metadata");
  const key = optionalString(request.key, "key");
  if (key === null) throw new Error("key must be a string");
  return {
    schema_version: TODO_COMPLETION_STATE_RESULT_SCHEMA,
    value: requireTodoCompletionMetadata(key, request.value),
  };
}

export function selectTodoCompletionContinuation(value: unknown): JsonObject {
  const request = requestObject(value, "continuation_for_write");
  return {
    schema_version: TODO_COMPLETION_STATE_RESULT_SCHEMA,
    continuation: completionContinuationForWrite(
      requiredBoolean(request.no_followup, "no_followup"),
      requiredBoolean(request.has_successor, "has_successor"),
    ),
  };
}

export function selectTodoCompletionState(value: unknown): TodoCompletionStateResult {
  const request = requestObject(value, "evaluate");
  const todo = requiredObject(request.todo, "todo");
  const requestedNoFollowup = requiredBoolean(
    request.requested_no_followup,
    "requested_no_followup",
  );
  const effectiveNoFollowup = requestedNoFollowup ||
    normalizeTodoNoFollowup(todo.no_followup) === true;
  const continuation = completionContinuationForWrite(
    effectiveNoFollowup,
    requiredBoolean(request.has_successor, "has_successor"),
  );
  const completionIdentitySource = optionalString(
    request.completion_identity_source,
    "completion_identity_source",
  );
  if (
    completionIdentitySource !== null &&
    completionIdentitySource !== "turn_settlement" &&
    completionIdentitySource !== "unscoped_completion" &&
    completionIdentitySource !== "lifecycle_reentry"
  ) {
    throw new Error("completion_identity_source is unsupported");
  }
  const recovery = String(todo.status ?? "") === "done" &&
      requestedNoFollowup &&
      normalizeTodoCompletionContinuation(todo.completion_continuation) ===
        "active_goal"
    ? completionIdentitySource === "lifecycle_reentry"
      ? "lifecycle_reentry_terminal_closeout"
      : "same_turn_terminal_closeout"
    : null;
  return {
    schema_version: TODO_COMPLETION_STATE_RESULT_SCHEMA,
    continuation,
    recovery,
  };
}

export function buildTodoCompletionMetadataUpdates(value: unknown): JsonObject {
  const request = requestObject(value, "metadata_updates");
  const block = requiredObject(request.block, "block");
  const completionContinuation = optionalString(
    request.completion_continuation,
    "completion_continuation",
  );
  const completionRecovery = optionalString(
    request.completion_recovery,
    "completion_recovery",
  );
  const updates: JsonObject = {};
  if (completionContinuation !== null) {
    updates.completion_continuation = completionContinuation;
  }
  if (completionRecovery !== null) updates.completion_recovery = completionRecovery;

  const targetStatus = optionalString(request.target_status, "target_status");
  const normalizedStatus = optionalString(
    request.normalized_status,
    "normalized_status",
  );
  if (
    targetStatus !== "done" || normalizedStatus !== "done" ||
    completionContinuation !== null ||
    normalizeTodoCompletionContinuation(block.completion_continuation) !== null
  ) {
    return {
      schema_version: TODO_COMPLETION_STATE_RESULT_SCHEMA,
      updates,
    };
  }

  const requestedNoFollowup = optionalBoolean(request.no_followup, "no_followup");
  const effectiveNoFollowup = requestedNoFollowup === null
    ? normalizeTodoNoFollowup(block.no_followup) === true
    : requestedNoFollowup;
  const successorTodoIds = request.successor_todo_ids === null
    ? stringArray(block.successor_todo_ids, "block.successor_todo_ids")
    : stringArray(request.successor_todo_ids, "successor_todo_ids");
  updates.completion_continuation = completionContinuationForWrite(
    effectiveNoFollowup,
    successorTodoIds.length > 0,
  );
  return {
    schema_version: TODO_COMPLETION_STATE_RESULT_SCHEMA,
    updates,
  };
}
